File: apps/dimi/src/tui/controllers/tasks-browser.ts
import type { BackgroundTaskInfo, Session } from '@dimi-agent/dimi-sdk';
import type { Component, ProcessTerminal, TUI } from '@dimi-agent/pi-tui';

import { TaskOutputViewer } from '../components/dialogs/task-output-viewer';
import { TasksBrowserApp, type TasksFilter } from '../components/dialogs/tasks-browser';
import type { Theme } from '#/tui/theme';
import type { CustomEditor } from '../components/editor/custom-editor';

export interface TasksBrowserHost {
  readonly state: {
    readonly tasksBrowser: TasksBrowserState | undefined;
    readonly theme: Theme;
    readonly terminal: ProcessTerminal;
    readonly ui: TUI;
    readonly editor: CustomEditor;
  };
  readonly backgroundTasks: ReadonlyMap<string, BackgroundTaskInfo>;
  readonly session: Session | undefined;
  showError(msg: string): void;
  setTasksBrowser(value: TasksBrowserState | undefined): void;
}

export type TasksBrowserState = {
  component: TasksBrowserApp;
  savedChildren: readonly Component[];
  filter: TasksFilter;
  selectedTaskId: string | undefined;
  tailOutput: string | undefined;
  tailLoading: boolean;
  tailRequestId: number;
  flashMessage: string | undefined;
  flashTimer: NodeJS.Timeout | undefined;
  pollTimer: NodeJS.Timeout | undefined;
  viewer:
    | {
        component: TaskOutputViewer;
        savedChildren: readonly Component[];
        taskId: string;
        output: string;
        refreshId: number;
        pollTimer: NodeJS.Timeout;
      }
    | undefined;
};

export class TasksBrowserController {
  constructor(private readonly host: TasksBrowserHost) {}

  async show(): Promise<void> {
    const { state } = this.host;
    if (state.tasksBrowser !== undefined) return;

    const session = this.host.session;
    if (session === undefined) {
      this.host.showError('No active session.');
      return;
    }

    let tasks: readonly BackgroundTaskInfo[] = [];
    try {
      tasks = await session.listBackgroundTasks({ activeOnly: false });
    } catch (error) {
      this.host.showError(
        `Failed to load tasks: ${error instanceof Error ? error.message : String(error)}`,
      );
      return;
    }
    if (state.tasksBrowser !== undefined) return;

    const filter: TasksFilter = 'all';
    const selectedTaskId = this.pickInitialSelection(tasks, filter);
    const component = new TasksBrowserApp(
      {
        tasks,
        filter,
        selectedTaskId,
        tailOutput: undefined,
        tailLoading: false,
        flashMessage: undefined,
        ...this.buildCallbacks(),
      },
      state.terminal,
    );

    const savedChildren = [...state.ui.children];
    state.ui.clear();
    state.ui.addChild(component);
    state.ui.setFocus(component);
    state.ui.requestRender(true);

    const pollTimer = setInterval(() => {
      void this.refresh({ silent: true });
    }, 1000);

    this.host.setTasksBrowser({
      component,
      savedChildren,
      filter,
      selectedTaskId,
      tailOutput: undefined,
      tailLoading: false,
      tailRequestId: 0,
      flashMessage: undefined,
      flashTimer: undefined,
      pollTimer,
      viewer: undefined,
    });

    if (selectedTaskId !== undefined) {
      this.loadTail(selectedTaskId);
    }
  }

  close(): void {
    const { state } = this.host;
    const browser = state.tasksBrowser;
    if (browser === undefined) return;
    if (browser.viewer !== undefined) this.closeOutputViewer();
    if (browser.pollTimer !== undefined) clearInterval(browser.pollTimer);
    if (browser.flashTimer !== undefined) clearTimeout(browser.flashTimer);

    state.ui.clear();
    for (const child of browser.savedChildren) {
      state.ui.addChild(child);
    }
    this.host.setTasksBrowser(undefined);
    state.ui.setFocus(state.editor);
    state.ui.requestRender(true);
  }

  repaint(): void {
    const browser = this.host.state.tasksBrowser;
    if (browser === undefined) return;
    const tasks = [...this.host.backgroundTasks.values()];
    this.pushProps(tasks);
  }

  async refreshOutputViewer(opts: { silent?: boolean } = {}): Promise<void> {
    const { state } = this.host;
    const browser = state.tasksBrowser;
    const viewer = browser?.viewer;
    if (browser === undefined || viewer === undefined) return;

    const session = this.host.session;
    if (session === undefined) return;

    const myRefreshId = ++viewer.refreshId;
    let output: string;
    try {
      output = await session.getBackgroundTaskOutput(viewer.taskId);
    } catch (error) {
      if (!opts.silent) {
        const message = error instanceof Error ? error.message : String(error);
        this.flash(`Output refresh failed: ${message}`);
      }
      return;
    }
    const current = state.tasksBrowser?.viewer;
    if (current === undefined || current !== viewer || current.refreshId !== myRefreshId) {
      return;
    }
    if (output === viewer.output) return;
    viewer.output = output;
    const info = this.host.backgroundTasks.get(viewer.taskId);
    viewer.component.setProps({
      taskId: viewer.taskId,
      info,
      output,
      onClose: () => {
        this.closeOutputViewer();
      },
    });
    state.ui.requestRender();
  }

  // ---------------------------------------------------------------------------

  private pickInitialSelection(
    tasks: readonly BackgroundTaskInfo[],
    filter: TasksFilter,
  ): string | undefined {
    const candidates =
      filter === 'all'
        ? tasks
        : tasks.filter(
            (t) =>
              t.status !== 'completed' &&
              t.status !== 'failed' &&
              t.status !== 'timed_out' &&
              t.status !== 'killed' &&
              t.status !== 'lost',
          );
    if (candidates.length === 0) return undefined;
    return candidates.find((t) => t.status === 'running')?.taskId ?? candidates[0]!.taskId;
  }

  private async refresh(opts: { silent?: boolean } = {}): Promise<void> {
    const { state } = this.host;
    const browser = state.tasksBrowser;
    if (browser === undefined) return;

    const session = this.host.session;
    if (session === undefined) return;

    let tasks: readonly BackgroundTaskInfo[];
    try {
      tasks = await session.listBackgroundTasks({ activeOnly: false });
    } catch (error) {
      if (!opts.silent) {
        this.flash(
          `Refresh failed: ${error instanceof Error ? error.message : String(error)}`,
        );
      }
      return;
    }
    if (state.tasksBrowser !== browser) return;
    this.pushProps(tasks);
  }

  private pushProps(tasks: readonly BackgroundTaskInfo[]): void {
    const browser = this.host.state.tasksBrowser;
    if (browser === undefined) return;
    browser.component.setProps({
      tasks,
      filter: browser.filter,
      selectedTaskId: browser.selectedTaskId,
      tailOutput: browser.tailOutput,
      tailLoading: browser.tailLoading,
      flashMessage: browser.flashMessage,
      ...this.buildCallbacks(),
    });
    this.host.state.ui.requestRender();
  }

  private buildCallbacks(): {
    onSelect: (taskId: string) => void;
    onToggleFilter: () => void;
    onRefresh: () => void;
    onCancel: () => void;
    onStopConfirmed: (taskId: string) => void;
    onOpenOutput: (taskId: string) => void;
    onStopIgnored: (taskId: string, reason: 'terminal') => void;
  } {
    return {
      onSelect: (taskId) => {
        this.handleSelect(taskId);
      },
      onToggleFilter: () => {
        this.handleToggleFilter();
      },
      onRefresh: () => {
        this.handleRefresh();
      },
      onCancel: () => {
        this.close();
      },
      onStopConfirmed: (taskId) => {
        void this.handleStop(taskId);
      },
      onOpenOutput: (taskId) => {
        void this.handleOpenOutput(taskId);
      },
      onStopIgnored: (taskId, reason) => {
        if (reason === 'terminal') {
          this.flash(`${taskId} is already terminal — nothing to stop.`);
        }
      },
    };
  }

  private handleSelect(taskId: string): void {
    const browser = this.host.state.tasksBrowser;
    if (browser === undefined) return;
    if (browser.selectedTaskId === taskId) return;
    browser.selectedTaskId = taskId;
    browser.tailOutput = undefined;
    browser.tailLoading = true;
    this.repaint();
    this.loadTail(taskId);
  }

  private handleToggleFilter(): void {
    const browser = this.host.state.tasksBrowser;
    if (browser === undefined) return;
    browser.filter = browser.filter === 'all' ? 'active' : 'all';
    this.repaint();
  }

  private handleRefresh(): void {
    this.flash('Refreshing…', 600);
    void this.refresh();
  }

  private async handleStop(taskId: string): Promise<void> {
    const browser = this.host.state.tasksBrowser;
    if (browser === undefined) return;

    const session = this.host.session;
    if (session === undefined) {
      this.flash('No active session.');
      return;
    }

    this.flash(`Stopping ${taskId}…`, 1500);
    try {
      await session.stopBackgroundTask(taskId, { reason: 'User initiated stop' });
      await this.refresh({ silent: true });
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      this.flash(`Stop failed: ${message}`);
    }
  }

  private async handleOpenOutput(taskId: string): Promise<void> {
    const { state } = this.host;
    const browser = state.tasksBrowser;
    if (browser === undefined) return;
    if (browser.viewer !== undefined) return;

    const session = this.host.session;
    if (session === undefined) {
      this.flash('No active session.');
      return;
    }

    let output: string;
    try {
      output = await session.getBackgroundTaskOutput(taskId);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      this.flash(`Cannot open output: ${message}`);
      return;
    }
    const current = state.tasksBrowser;
    if (current === undefined || current !== browser) return;

    const info = this.host.backgroundTasks.get(taskId);
    const viewer = new TaskOutputViewer(
      {
        taskId,
        info,
        output,
        onClose: () => {
          this.closeOutputViewer();
        },
      },
      state.terminal,
    );

    const savedBrowserChildren = [...state.ui.children];
    state.ui.clear();
    state.ui.addChild(viewer);
    state.ui.setFocus(viewer);
    state.ui.requestRender(true);

    const pollTimer = setInterval(() => {
      void this.refreshOutputViewer({ silent: true });
    }, 1000);

    browser.viewer = {
      component: viewer,
      savedChildren: savedBrowserChildren,
      taskId,
      output,
      refreshId: 0,
      pollTimer,
    };
  }

  private loadTail(taskId: string): void {
    const { state } = this.host;
    const browser = state.tasksBrowser;
    if (browser === undefined) return;

    const session = this.host.session;
    if (session === undefined) {
      browser.tailLoading = false;
      this.repaint();
      return;
    }

    const requestId = ++browser.tailRequestId;
    void session
      .getBackgroundTaskOutput(taskId, { tail: 4000 })
      .then((output) => {
        const current = state.tasksBrowser;
        if (current === undefined) return;
        if (current !== browser || current.tailRequestId !== requestId) return;
        if (current.selectedTaskId !== taskId) return;
        current.tailOutput = output;
        current.tailLoading = false;
        this.repaint();
      })
      .catch(() => {
        const current = state.tasksBrowser;
        if (current === undefined) return;
        if (current !== browser || current.tailRequestId !== requestId) return;
        if (current.selectedTaskId !== taskId) return;
        current.tailOutput = '';
        current.tailLoading = false;
        this.repaint();
      });
  }

  private flash(message: string, durationMs = 2500): void {
    const browser = this.host.state.tasksBrowser;
    if (browser === undefined) return;
    if (browser.flashTimer !== undefined) clearTimeout(browser.flashTimer);
    browser.flashMessage = message;
    browser.flashTimer = setTimeout(() => {
      const current = this.host.state.tasksBrowser;
      if (current !== browser) return;
      current.flashMessage = undefined;
      current.flashTimer = undefined;
      this.repaint();
    }, durationMs);
    this.repaint();
  }

  private closeOutputViewer(): void {
    const browser = this.host.state.tasksBrowser;
    if (browser === undefined || browser.viewer === undefined) return;
    const viewer = browser.viewer;
    clearInterval(viewer.pollTimer);
    browser.viewer = undefined;
    this.host.state.ui.clear();
    for (const child of viewer.savedChildren) {
      this.host.state.ui.addChild(child);
    }
    this.host.state.ui.setFocus(browser.component);
    this.host.state.ui.requestRender(true);
  }
}

File: crates/dimi-tui/src/controllers/tasks_browser.rs
//! Tasks browser controller — the state-machine part of
//! `apps/dimi/src/tui/controllers/tasks-browser.ts`.
//!
//! Owns the task list state (filter all/active, selection, tail output,
//! flash messages, output viewer) and the pure transitions
//! (`handleSelect`, `handleToggleFilter`, `handleStop`, `loadTail`,
//! `handleOpenOutput`, …). The `TasksBrowserApp` component, the `/tasks`
//! polling timers, and the async `session.listBackgroundTasks` /
//! `stopBackgroundTask` / `getBackgroundTaskOutput` calls are
//! `// TODO(legacy)`.

use crate::controllers::{BackgroundTaskInfo, BackgroundTaskKind, BackgroundTaskStatus};

/// `TasksFilter` — the `all` / `active` toggle.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TasksFilter {
    All,
    Active,
}

impl TasksFilter {
    pub fn toggled(&self) -> TasksFilter {
        match self {
            TasksFilter::All => TasksFilter::Active,
            TasksFilter::Active => TasksFilter::All,
        }
    }
}

/// Output-viewer sub-state (`TasksBrowserState.viewer`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskViewerState {
    pub task_id: String,
    pub output: String,
    pub refresh_id: u64,
}

/// The tasks-browser state (`TasksBrowserState`), minus the component/timers.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TasksBrowserState {
    pub open: bool,
    pub filter: TasksFilter,
    pub selected_task_id: Option<String>,
    pub tail_output: Option<String>,
    pub tail_loading: bool,
    pub tail_request_id: u64,
    pub flash_message: Option<String>,
    pub viewer: Option<TaskViewerState>,
}

impl Default for TasksBrowserState {
    fn default() -> Self {
        TasksBrowserState {
            open: false,
            filter: TasksFilter::All,
            selected_task_id: None,
            tail_output: None,
            tail_loading: false,
            tail_request_id: 0,
            flash_message: None,
            viewer: None,
        }
    }
}

/// Result of [`TasksBrowserController::show`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ShowOutcome {
    /// The browser is already open.
    AlreadyOpen,
    /// No active session to list tasks from.
    NoSession,
    /// Opened; the selected task (if any) has a tail load pending.
    Opened { selected: Option<String> },
}

/// The tasks browser controller (port of `TasksBrowserController`).
#[derive(Debug, Default)]
pub struct TasksBrowserController {
    pub state: TasksBrowserState,
}

impl TasksBrowserController {
    pub fn new() -> Self {
        TasksBrowserController::default()
    }

    /// `pickInitialSelection` — prefer a `running` task, else the first
    /// candidate (filtered to active when the filter is `active`).
    pub fn pick_initial_selection(
        &self,
        tasks: &[BackgroundTaskInfo],
        filter: TasksFilter,
    ) -> Option<String> {
        let candidates: Vec<&BackgroundTaskInfo> = match filter {
            TasksFilter::All => tasks.iter().collect(),
            TasksFilter::Active => tasks.iter().filter(|t| !t.status.is_terminal()).collect(),
        };
        if candidates.is_empty() {
            return None;
        }
        candidates
            .iter()
            .find(|t| t.status == BackgroundTaskStatus::Running)
            .map(|t| t.task_id.clone())
            .or_else(|| Some(candidates[0].task_id.clone()))
    }

    /// `show`.
    pub fn show(&mut self, tasks: &[BackgroundTaskInfo], has_session: bool) -> ShowOutcome {
        if self.state.open {
            return ShowOutcome::AlreadyOpen;
        }
        if !has_session {
            // TODO(legacy): host.showError('No active session.')
            self.state.flash_message = Some("No active session.".to_owned());
            return ShowOutcome::NoSession;
        }
        // TODO(legacy): async session.listBackgroundTasks({ activeOnly: false })
        let filter = TasksFilter::All;
        let selected = self.pick_initial_selection(tasks, filter);
        self.state.open = true;
        self.state.filter = filter;
        self.state.selected_task_id = selected.clone();
        self.state.tail_output = None;
        self.state.tail_loading = false;
        self.state.tail_request_id = 0;
        self.state.flash_message = None;
        self.state.viewer = None;
        // TODO(legacy): create TasksBrowserApp + mount; start 1s poll timer
        if let Some(selected) = &selected {
            self.load_tail(selected, has_session);
        }
        ShowOutcome::Opened { selected }
    }

    /// `close`.
    pub fn close(&mut self) {
        if !self.state.open {
            return;
        }
        if self.state.viewer.is_some() {
            self.close_output_viewer();
        }
        // TODO(legacy): clear pollTimer / flashTimer; restore saved children
        self.state.open = false;
        self.state.viewer = None;
    }

    /// `repaint` — the prop push to `TasksBrowserApp` is legacy; the state is
    /// already up to date.
    pub fn repaint(&mut self) {
        // TODO(legacy): browser.component.setProps({ tasks, filter, … })
    }

    /// `handleSelect`.
    pub fn handle_select(&mut self, task_id: &str) {
        if !self.state.open {
            return;
        }
        if self.state.selected_task_id.as_deref() == Some(task_id) {
            return;
        }
        self.state.selected_task_id = Some(task_id.to_owned());
        self.state.tail_output = None;
        self.state.tail_loading = true;
        self.repaint();
        self.load_tail(task_id, true);
    }

    /// `handleToggleFilter`.
    pub fn handle_toggle_filter(&mut self) {
        if !self.state.open {
            return;
        }
        self.state.filter = self.state.filter.toggled();
        self.repaint();
    }

    /// `handleRefresh` — flash "Refreshing…" and kick a (legacy) refresh.
    pub fn handle_refresh(&mut self) {
        if !self.state.open {
            return;
        }
        self.flash("Refreshing…", 600);
        // TODO(legacy): void this.refresh() (async listBackgroundTasks → pushProps)
    }

    /// `handleStop` — flash the stopping state; the actual
    /// `session.stopBackgroundTask` is legacy. Success/failure land via
    /// [`TasksBrowserController::handle_stop_result`].
    pub fn handle_stop(&mut self, task_id: &str, has_session: bool) {
        if !self.state.open {
            return;
        }
        if !has_session {
            self.flash("No active session.", 2500);
            return;
        }
        self.flash(&format!("Stopping {task_id}…"), 1500);
        // TODO(legacy): async session.stopBackgroundTask(taskId) → refresh
    }

    /// Surface a stop failure (`Stop failed: …`).
    pub fn handle_stop_failure(&mut self, message: &str) {
        if !self.state.open {
            return;
        }
        self.flash(&format!("Stop failed: {message}"), 2500);
    }

    /// `handleOpenOutput` — the fetch is legacy; the host reports the output
    /// via [`TasksBrowserController::open_output_result`].
    pub fn handle_open_output(&mut self, task_id: &str, has_session: bool) {
        if !self.state.open {
            return;
        }
        if self.state.viewer.is_some() {
            return;
        }
        if !has_session {
            self.flash("No active session.", 2500);
            return;
        }
        // TODO(legacy): async session.getBackgroundTaskOutput(taskId)
        let _ = task_id;
    }

    /// Complete `handleOpenOutput` with the fetched output (or an error).
    pub fn open_output_result(&mut self, task_id: &str, output: Result<String, String>) {
        if !self.state.open {
            return;
        }
        match output {
            Ok(output) => {
                self.state.viewer = Some(TaskViewerState {
                    task_id: task_id.to_owned(),
                    output,
                    refresh_id: 0,
                });
                // TODO(legacy): create TaskOutputViewer + mount; start 1s poll
            }
            Err(message) => {
                self.flash(&format!("Cannot open output: {message}"), 2500);
            }
        }
    }

    /// `loadTail` — bumps the request id; the fetch is legacy.
    pub fn load_tail(&mut self, task_id: &str, has_session: bool) {
        if !self.state.open {
            return;
        }
        if !has_session {
            self.state.tail_loading = false;
            self.repaint();
            return;
        }
        self.state.tail_request_id += 1;
        // TODO(legacy): async session.getBackgroundTaskOutput(taskId, { tail: 4000 })
        let _ = task_id;
    }

    /// Settle a tail fetch. `output` is `None` on error (→ empty output).
    pub fn load_tail_result(&mut self, task_id: &str, request_id: u64, output: Option<String>) {
        if !self.state.open {
            return;
        }
        if self.state.tail_request_id != request_id {
            return;
        }
        if self.state.selected_task_id.as_deref() != Some(task_id) {
            return;
        }
        self.state.tail_output = Some(output.unwrap_or_default());
        self.state.tail_loading = false;
        self.repaint();
    }

    /// `flash`.
    pub fn flash(&mut self, message: &str, duration_ms: u64) {
        if !self.state.open {
            return;
        }
        // TODO(legacy): clearTimeout(flashTimer)
        self.state.flash_message = Some(message.to_owned());
        // TODO(legacy): flashTimer = setTimeout(() => { clearFlash() }, duration_ms)
        let _ = duration_ms;
    }

    /// The timer-expiry half of `flash` — clears the message.
    pub fn clear_flash(&mut self) {
        if !self.state.open {
            return;
        }
        self.state.flash_message = None;
        self.repaint();
    }

    /// `closeOutputViewer`.
    pub fn close_output_viewer(&mut self) {
        if self.state.viewer.is_none() {
            return;
        }
        // TODO(legacy): clearInterval(viewer.pollTimer); restore saved children
        self.state.viewer = None;
    }
}

/// `shouldShowBackgroundTaskTranscript`-independent terminal check helper.
pub fn is_terminal_status(status: BackgroundTaskStatus) -> bool {
    status.is_terminal()
}

/// Group helper used by the browser: count active (non-terminal) tasks.
pub fn count_active_tasks(tasks: &[BackgroundTaskInfo]) -> (usize, usize) {
    let mut bash = 0usize;
    let mut agents = 0usize;
    for task in tasks {
        if task.status.is_terminal() {
            continue;
        }
        if task.kind == BackgroundTaskKind::Agent {
            agents += 1;
        } else {
            bash += 1;
        }
    }
    (bash, agents)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn task(id: &str, status: BackgroundTaskStatus) -> BackgroundTaskInfo {
        BackgroundTaskInfo::new(id, BackgroundTaskKind::Process, status)
    }

    #[test]
    fn pick_initial_selection_prefers_running() {
        let c = TasksBrowserController::new();
        let tasks = vec![
            task("a", BackgroundTaskStatus::Completed),
            task("b", BackgroundTaskStatus::Running),
            task("c", BackgroundTaskStatus::Failed),
        ];
        assert_eq!(
            c.pick_initial_selection(&tasks, TasksFilter::All)
                .as_deref(),
            Some("b")
        );
        // Active filter drops terminal tasks.
        assert_eq!(
            c.pick_initial_selection(&tasks, TasksFilter::Active)
                .as_deref(),
            Some("b")
        );
    }

    #[test]
    fn pick_initial_selection_falls_back_to_first() {
        let c = TasksBrowserController::new();
        let tasks = vec![
            task("a", BackgroundTaskStatus::Running),
            task("z", BackgroundTaskStatus::Completed),
        ];
        assert_eq!(
            c.pick_initial_selection(&tasks, TasksFilter::All)
                .as_deref(),
            Some("a")
        );
        assert_eq!(c.pick_initial_selection(&[], TasksFilter::All), None);
    }

    #[test]
    fn show_requires_session_and_opens() {
        let mut c = TasksBrowserController::new();
        let tasks = vec![task("a", BackgroundTaskStatus::Running)];
        assert_eq!(c.show(&tasks, false), ShowOutcome::NoSession);
        assert!(!c.state.open);

        assert_eq!(
            c.show(&tasks, true),
            ShowOutcome::Opened {
                selected: Some("a".to_owned())
            }
        );
        assert!(c.state.open);
        // loadTail bumped the request id but leaves tail_loading false until
        // the async fetch settles.
        assert!(!c.state.tail_loading);
        assert_eq!(c.state.tail_request_id, 1);

        // Already open.
        assert_eq!(c.show(&tasks, true), ShowOutcome::AlreadyOpen);
    }

    #[test]
    fn select_toggles_tail_state() {
        let mut c = TasksBrowserController::new();
        c.show(&[task("a", BackgroundTaskStatus::Running)], true);
        c.state.tail_loading = false;

        c.handle_select("a");
        // Same selection → no change.
        assert!(!c.state.tail_loading);

        c.handle_select("b");
        assert_eq!(c.state.selected_task_id.as_deref(), Some("b"));
        assert!(c.state.tail_loading);
        assert!(c.state.tail_output.is_none());
    }

    #[test]
    fn filter_toggle_switches_all_active() {
        let mut c = TasksBrowserController::new();
        c.show(&[], true);
        assert_eq!(c.state.filter, TasksFilter::All);
        c.handle_toggle_filter();
        assert_eq!(c.state.filter, TasksFilter::Active);
        c.handle_toggle_filter();
        assert_eq!(c.state.filter, TasksFilter::All);
    }

    #[test]
    fn tail_request_id_guards_stale_results() {
        let mut c = TasksBrowserController::new();
        c.show(&[task("a", BackgroundTaskStatus::Running)], true);
        let req = c.state.tail_request_id;
        // A stale (older) request settles as a no-op.
        c.load_tail_result("a", req.saturating_sub(1), Some("stale".to_owned()));
        assert!(!c.state.tail_loading); // request already consumed by show? see below

        // Fresh request lands.
        c.load_tail("a", true);
        let req = c.state.tail_request_id;
        c.load_tail_result("a", req, Some("fresh output".to_owned()));
        assert_eq!(c.state.tail_output.as_deref(), Some("fresh output"));
        assert!(!c.state.tail_loading);
    }

    #[test]
    fn stop_and_stop_failure_flash() {
        let mut c = TasksBrowserController::new();
        c.show(&[], true);
        c.handle_stop("t1", true);
        assert_eq!(c.state.flash_message.as_deref(), Some("Stopping t1…"));
        c.handle_stop("t1", false);
        assert_eq!(c.state.flash_message.as_deref(), Some("No active session."));
        c.handle_stop_failure("boom");
        assert_eq!(c.state.flash_message.as_deref(), Some("Stop failed: boom"));
        c.clear_flash();
        assert!(c.state.flash_message.is_none());
    }

    #[test]
    fn open_output_viewer_and_close() {
        let mut c = TasksBrowserController::new();
        c.show(&[], true);
        c.open_output_result("t1", Ok("out".to_owned()));
        let viewer = c.state.viewer.as_ref().unwrap();
        assert_eq!(viewer.task_id, "t1");
        assert_eq!(viewer.output, "out");

        // Can't open a second viewer while one is open.
        c.handle_open_output("t2", true);
        assert_eq!(c.state.viewer.as_ref().unwrap().task_id, "t1");

        c.close_output_viewer();
        assert!(c.state.viewer.is_none());
    }

    #[test]
    fn open_output_error_flashes() {
        let mut c = TasksBrowserController::new();
        c.show(&[], true);
        c.open_output_result("t1", Err("denied".to_owned()));
        assert!(c.state.viewer.is_none());
        assert_eq!(
            c.state.flash_message.as_deref(),
            Some("Cannot open output: denied")
        );
    }

    #[test]
    fn count_active_tasks_ignores_terminal() {
        let tasks = vec![
            task("a", BackgroundTaskStatus::Running),
            task("b", BackgroundTaskStatus::Completed),
        ];
        assert_eq!(count_active_tasks(&tasks), (1, 0));
    }
}
